Miscellaneous refactor

erikkemperman opened the pull request:
I thought there was some room for improvement here, hope you don't mind my tacking these onto a single PR. In particular, the last change (exporting `isConstant`) would be needed to get the effort around node-glob caches working once again.

Separately I have some work which would prevent a bunch of spurious function binding, but that would depend on [this PR](https://github.com/gulpjs/value-or-function/pull/7) on `value-or-function` -- or something like it -- being approved and merged first.

For what it's worth though, I have it here:
https://github.com/erikkemperman/resolve-options/commits/refactor-misc-bind

erikkemperman committed:
Refactor: Test key argument type implicitly

erikkemperman committed:
Refactor: Failing test, no guard against indirect recursion

erikkemperman committed:
Refactor: Fix test, add guard against indirect recursion

erikkemperman committed:
Refactor: Pre-process options and fallbacks

phated commented:
@erikkemperman why should we export `isConstant` when it is used in the resolve?

erikkemperman commented:
@phated So we can decide whether to call `resolve` just once, up front, or repeatedly within an iteration over subjects -- depending on if a particular option is set to a constant (and thus will be same for all subjects).

I had a particular use case in vinyl-fs, whilst repairing my node-glob cache branch after the option resolver was adopted (ha):
https://github.com/erikkemperman/vinyl-fs/commit/45ca5ff3eb042982e4e29d94cd4fd1bedd1ae01a#diff-6fab5f49d036c0eb8350c49275b632acR12

But I think it might be generally useful. E.g. to avoid potentially expensive iterations:

```js
var subjects = ['my', 'huge', 'array'];
if (resolver.isConstant('keep')) {
  if (!resolver.resolve('keep')) {
    // constant false: clear all subjects
    subjects = []
  } else {
    // constant true: keep all subjects
  }
} else {
  // variable: evaluate per subject
  subjects = subjects.filter(function(subject) {
    return resolver.resolve('keep', subject);
  });
}
```

erikkemperman commented:
@phated I'll wait for the `value-or-function` changes to settle, will then amend this accordingly.

phated commented:
@erikkemperman but if the `isConstant` method is being used at the beginning of `resolve`, it will just bail early and the overhead of that function call is almost non-existent.  I also think we need to give v8 some benefit of the doubt that it will optimize/inline the result of constant calls.

To be clear, I like the `isConstant` change as an **internal** change but not exposed in the external API.  Thoughts?

phated commented:
`value-or-function` changes have been published as 3.0

erikkemperman committed:
Update: Bump value-or-function to 3.0.0

erikkemperman commented:
@phated True, but my thinking was that there is no sense in repeating that call over and over, even if the overhead is small, if we already know it will always yield the same value. Anyway, I will update this PR on top of value-or-function 3.0 and leave `isConstant()` out of it. Maybe I can get the node-glob cache branch to work without that.

erikkemperman commented:
Ah yes, now I remember -- I am not allowed to do `resolver.resolve('resolveSymlinks', file)` without first attaching symbolic stats to the file -- you added a [test](https://github.com/gulpjs/vinyl-fs/blob/master/test/src-symlinks.js#L130) for that explicitly.

But of course the whole point of leveraging the stat cache is that we can often avoid invoking `fs.lstat` in the first place, when we find (regular) stats cached and the `resolveSymlinks` option is true.

So I figured I would make that work at least for the (likely very common) case that `resolveSymlinks` is left to the default (constant) `true` -- that way your test condition holds up for scenarios where this option is given as a function, and we can still leverage the cache a lot of the time.

Alternatively, the requirement of having stats attached in all option callbacks could have an exception in `resolveSymlinks` -- but I think you already decided against that.

Apologies, I'm finding it hard to describe this properly, maybe better to point at some code:
https://github.com/erikkemperman/vinyl-fs/commits/use-cache

erikkemperman commented:
@phated Just a polite bump in case this has understandably fallen off your radar :-) As far as I can tell I've removed your main objection here, or at least the one you mentioned thus far, and I'd propose we deal separately with the use case I see for something like `isConstant`. Make sense?

phated commented:
@erikkemperman this totally got buried in all the noise, thanks for bumping.

phated commented on index.js:
```diff
@@ -14,47 +13,90 @@ function createResolver(config, options) {
     }
 
-    if (key === lastKey) {
+    if (stack.some(function(s) {
```
Make this a named function in the outermost possible scope please.

erikkemperman replied:
All right

phated commented on index.js:
```diff
@@ -14,47 +13,90 @@ function createResolver(config, options) {
-    if (typeof option === 'function') {
-      option = option.bind(resolver);
+    stack.push(key);
+    try {
```
I'd prefer not to have a try/catch.  Why was this added?

erikkemperman replied:
We're potentially calling user-defined functions in here, if those throw an error our stack would not get cleaned up.

phated commented on index.js:
```diff
@@ -14,47 +13,90 @@ function createResolver(config, options) {
+        option = undefined;
+      }
+
+      if (option === undefined) {
```
Shouldn't this be checking against `== null` to check null or undefined?

erikkemperman replied:
Ah yes, yes it should. Leftover from my suggestion earlier about having value-or-function use strictly undefined for rejection.

phated commented on index.js:
```diff
@@ -14,47 +13,90 @@ function createResolver(config, options) {
+      stack.pop();
+      return option;
+
+    } catch (err) {
```
If we **must** keep the `try`, this should probably be a finally, no?

erikkemperman replied:
I figured that we would not want to swallow any errors here, so I rethrow after cleaning up the stack.

erikkemperman replied:
Shouldn't post before first coffee, sorry. Yes `finally` would work and won't swallow any errors. Will amend.

phated commented on index.js:
```diff
@@ -14,47 +13,90 @@ function createResolver(config, options) {
-    var args = [definition.type, option].concat(appliedArgs);
-    var result = normalize.apply(null, args);
+  // Pre-process
+  options = Object.keys(config).reduce(function(opts, key) {
```
is there a reason to pre-process?  I think if we just cache a constant on the first access, that would be more performant in the cases that not all options are always accessed.

erikkemperman replied:
Hmm, I see what you mean, but that would perhaps close the door on adding something like `isConstant` and allowing that to be called without resolving anything.

So maybe my idea of discussing that separately was too simplistic... But let me try if I can come up with something in between, sort of.

phated reviewed (changes requested):
@erikkemperman a few things I noticed.

erikkemperman committed:
Amend: address review

erikkemperman commented:
Thanks for reviewing this! Pushed some changes, addressing most of what you found. The pre-processing is still in there, I haven't had a lot of time to look at it but believe it will be tricky/ugly to get rid of it without definitively shutting avenues to using node-glob stat cache in vinyl-fs... At least not if, as I presume, you will insist upon having symbolic stats attached to each vinyl argument to `resolveSymlinks` callbacks, as in your test (linked above).

Personally, I think it's worth a bit of ugliness / overhead, whether through an `isConstant` export or some other way, because of the much larger performance penalty of doing up to two spurious `fs.(l)stat` calls which can be avoided in many common scenarios using node-glob's cache.

erikkemperman committed:
Update: Add resolveConstants export

erikkemperman commented:
Just to throw something concrete out there, I managed to avoid pre-processing anything (still caching constants but only when the key is actually requested) whilst still making it possible to check for constant options without invoking function options.

I did so by exporting a separate `resolveConstant` method, which is then used by the main `resolve` method as well (this on top of the present PR):
https://github.com/erikkemperman/resolve-options/commits/refactor-misc-constant

I've tentatively updated my use-cache branch of vinyl-fs to demonstrate the use case:
https://github.com/erikkemperman/vinyl-fs/commits/use-cache

And the general case I included above remains valid, but slightly modified:

```js
var subjects = ['my', 'huge', 'array'];
var keep = resolver.resolveConstant('keep');

if (typeof keep === 'boolean') {
  if (!resolver.resolve('keep')) {
    // constant false: clear all subjects
    subjects = []

  } else {
    // constant true: keep all subjects
  }

} else {
  // variable: evaluate per subject
  subjects = subjects.filter(function(subject) {
    return resolver.resolve('keep', subject);
  });
}
```

BTW I was thinking I could actually treat zero-arg functions the same as constants, after invoking them once, immediately up front -- does that sound reasonable?

phated commented on index.js:
```diff
@@ -14,47 +13,86 @@ function createResolver(config, options) {
     }
 
-    if (key === lastKey) {
+    if (stack.indexOf(key) >= 0) {
```
👍

phated commented:
@erikkemperman I hate to keep pushing back on your `use-cache` PR but I'm trying to take all of this in the smallest steps possible.  I like your `resolveConstant` changes but don't want to export the method to the public API in this iteration.  Can you add those changes to this PR without the export?

phated commented on index.js:
```diff
@@ -14,47 +13,86 @@ function createResolver(config, options) {
-    if (typeof option === 'function') {
-      option = option.bind(resolver);
+    stack.push(key);
+    try {
```
This should be abstracted as noted in https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax - try/finally wasn't optimizable until the V8 in node 7.

erikkemperman committed:
Amend: address review

erikkemperman committed:
Amend: address review

erikkemperman commented:
@phated No worries, I appreciate your taking the time for this stuff and am learning new things which is always nice. Like the fact that V8 didn't optimize try/catch until so recently!

I think I've addressed your remarks. Done so in additive commits, to not mess with the review. But (wishful thinking perhaps) it feels like this might be nearing merge so took the liberty of creating a separate branch in which I reordered and squashed history in a few topical commits:

https://github.com/erikkemperman/resolve-options/commits/refactor-misc-clean

The diff against master is identical to this PR.

I've separated out the export of `resolveConstant`, but in case we want that at some point I've made a branch, on top of the clean one liked above:

https://github.com/erikkemperman/resolve-options/commits/export-resolve-constants

Thanks again!

phated commented on index.js:
```diff
@@ -14,44 +13,89 @@ function createResolver(config, options) {
-    var appliedArgs = slice.call(arguments, 1);
+    var option = options[key];
+
+    if (!!option || options.hasOwnProperty(key)) {
```
What is this checking?  Could we just check `option == null`?

erikkemperman replied:
It'd have to be `option != null`, if I understand you right, and would not catch `false` as an explicitly set value, when it should be treated as a constant... The thought of this test was to catch the common case of a truthy value without further function call and to catch explicit falsy values with `hasOwnProperty`. Make sense?

phated commented on index.js:
```diff
@@ -14,44 +13,89 @@ function createResolver(config, options) {
+    var option = options[key];
+
+    if (!!option || options.hasOwnProperty(key)) {
+      if (typeof option !== 'function') {
```
If you invert this, you can just return early and reduce a bunch of this indentation.  It would also group things nicely at the bottom of the function and avoid your `// Fall through` scenario.  If you simplify the above line, you could group the early return into a single `if` statement.

erikkemperman replied:
Agree about the early return, but I don't think I can merge the `if`s?

phated commented on index.js:
```diff
@@ -14,44 +13,89 @@ function createResolver(config, options) {
-    return koalas(result, fallback);
+  function tryResolve(option, fallback, appliedArgs, args) {
+    try {
+      option = normalize.apply(resolver, args);
```
I believe the body of this `try` should be place inside a function (probably name it `toResolve` or something?) inside the `resolve` method and then passed as the only argument to `tryResolve`.  Maybe `stack.pop` should be passed as a 2nd argument, so it can be made a top-level utility method instead of existing in this closure.

erikkemperman replied:
I think I just did what you requested, but actually I believe the performance hit of try/catch was already averted? Still, this newer version has to pass around some fewer args, so that's probably for the best.

phated reviewed (changes requested):
@erikkemperman this PR is looking great, just a few minor cleanup things remain.

erikkemperman committed:
Amend: address review

erikkemperman commented:
@phated Amended, also the other branches I mentioned.

erikkemperman committed:
Amend: address review

erikkemperman commented:
@phated Just occurred to me I completely misread your comment [here](https://github.com/gulpjs/resolve-options/pull/1#discussion_r131532150), disregard my earlier reply please :-) Simplified the test just now, other branches amended as well.

phated commented:
@erikkemperman woot! After a really quick review, this looks great!  I might do some more cleanup after a full review but I'll get this merged before you get back.

erikkemperman commented:
Cool! Just to reiterate, I linked a "clean" branch above which has the same changes but in just a few topical commits. Have a great weekend!

phated commented:
@erikkemperman I swear I've tried to get this done 2-3 times since the last comments and keep getting pulled away.  I'll try to get it done today (though it's almost happy hour).

phated commented:
Merged as a fast-forward from your `refactor-misc-clean` branch.  @erikkemperman I just wanted to say that this PR was amazing (this was the first time I reviewed the additional test cases).  I'm also very happy with my initial abstraction because you were able to easy add a big performance improvement without changing the API at all.

erikkemperman commented:
@phated Thanks a lot for reviewing this so thoroughly, it's improved a lot since the first attempt!
